lib/crypto-primitives/src/Crypto/Encryption/ChaChaPoly1305.hs

```diff
@@ -0,0 +1,93 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+
+module Crypto.Encryption.ChaChaPoly1305
+    ( encryptPayload
+    , decryptPayload
+    , toSymmetricKey
+
+    , CryptoError (..)
+    , CryptoFailable (..)
+    ) where
+
+import Prelude
+
+import Control.Monad
+    ( when
+    )
+import Crypto.Error
+    ( CryptoError (..)
+    , CryptoFailable (..)
+    )
+import Crypto.Hash.Algorithms
+    ( SHA512 (..)
+    )
+import Data.ByteString
+    ( ByteString
+    )
+
+import qualified Codec.CBOR.Encoding as CBOR
+import qualified Codec.CBOR.Write as CBOR
+import qualified Crypto.Cipher.ChaChaPoly1305 as Poly
+import qualified Crypto.KDF.PBKDF2 as PBKDF2
+import qualified Data.ByteArray as BA
+import qualified Data.ByteString as BS
+
+-- | ChaCha20/Poly1305 encrypting the payload with passphrase and nonce.
+-- | The caller must ensure that the passphrase length is 32 bytes and
+-- | nonce is 12 bytes
+encryptPayload
+    :: ByteString
+       -- ^ Symmetric key / passphrase, 32-byte long
+    -> ByteString
+       -- ^ Nonce, 12-byte long
+    -> ByteString
+        -- ^ Payload to be encrypted
```

Comment:
I suggest to use newtypes or `Tagged` to distinguish parameters on the type level.